pkg/cgroups/cgroupblkio.go

```diff
@@ -0,0 +1,188 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cgroups
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+
+	logger "github.com/intel/cri-resource-manager/pkg/log"
+	_ "github.com/intel/cri-resource-manager/pkg/sysfs"
+	"github.com/intel/cri-resource-manager/pkg/utils"
+)
+
+const (
+	blkioCgroupDir = "/sys/fs/cgroup/blkio/"
+)
+
+// cgroups blkio parameter filenames.
+var blkioWeightFiles = []string{"blkio.weight", "blkio.bfq.weight", "blkio.cfq.weight"}
+var blkioLeafWeightFiles = []string{"blkio.leaf_weight", "blkio.bfq.leaf_weight", "blkio.cfq.leaf_weight"}
+var blkioWeightDeviceFiles = []string{"blkio.weight_device", "blkio.bfq.weight_device", "blkio.cfq.weight_device"}
+var blkioThrottleReadBpsFiles = []string{"blkio.throttle.read_bps_device"}
+var blkioThrottleWriteBpsFiles = []string{"blkio.throttle.write_bps_device"}
+var blkioThrottleReadIOPSFiles = []string{"blkio.throttle.read_iops_device"}
+var blkioThrottleWriteIOPSFiles = []string{"blkio.throttle.write_iops_device"}
+
+// OciBlockIOParameters contains OCI standard configuration of cgroups blkio parameters.
+//
+// Effects of Weight, LeafWeight and Rate values in SetBlkioParameters():
+// Value  |  Effect
+// -------+-------------------------------------------------------------------
+//    -1  |  Do not write to cgroups, value is missing
+//     0  |  Write to cgroups, will remove the setting as specified in cgroups blkio interface
+//  other |  Write to cgroups, sets the value
+type OciBlockIOParameters struct {
+	Weight                  int64
+	LeafWeight              int64
+	WeightDevice            []OciWeightDeviceParameters
+	ThrottleReadBpsDevice   []OciRateDeviceParameters
+	ThrottleWriteBpsDevice  []OciRateDeviceParameters
+	ThrottleReadIOPSDevice  []OciRateDeviceParameters
+	ThrottleWriteIOPSDevice []OciRateDeviceParameters
+}
+
+// OciWeightDeviceParameters contains values for
+// - blkio.[io-scheduler].weight
+// - blkio.[io-scheduler].leaf_weight
+type OciWeightDeviceParameters struct {
+	Major      int64
+	Minor      int64
+	Weight     int64
+	LeafWeight int64
+}
+
+// OciRateDeviceParameters contains values for
+// - blkio.throttle.read_bps_device
+// - blkio.throttle.write_bps_device
+// - blkio.throttle.read_iops_device
+// - blkio.throttle.write_iops_device
+type OciRateDeviceParameters struct {
+	Major int64
+	Minor int64
+	Rate  int64
+}
+
+// NewOciBlockIOParameters creates new OciBlockIOParameters instance.
+func NewOciBlockIOParameters() OciBlockIOParameters {
+	return OciBlockIOParameters{
+		Weight:                  -1,
+		LeafWeight:              -1,
+		WeightDevice:            nil,
+		ThrottleReadBpsDevice:   nil,
+		ThrottleWriteBpsDevice:  nil,
+		ThrottleReadIOPSDevice:  nil,
+		ThrottleWriteIOPSDevice: nil,
```

Comment:
nitpick: golang is actively discouraging labelled initialization of composite members to zero value. IOW, you should not set anything else than those -1's above.